tools/cloud/download.xml

```diff
@@ -0,0 +1,86 @@
+<?xml version="1.0"?>
+
+<tool id="download_to_cloud" name="Download dataset" version="0.1.0" workflow_compatible="false">
+    <description>
+        to Cloud
+    </description>
+    <requirements>
+        <requirement type="package" version="0.3.3">cloudbridge</requirement>
+    </requirements>
+    <command interpreter="python">
+        download.py
+            --filename=$filename
+            --provider=$connection.provider
+            #if $connection.provider=="aws"
+                --ca_access_key=$connection.access_key
+                --ca_secret_key=$connection.secret_key
+            #else if $connection.provider=="azure"
+                --cm_subscription_id=$connection.subscription_id
+                --cm_client_id=$connection.client_id
+                --cm_secret=$connection.secret
+                --cm_tenant=$connection.tenant
+                --cm_storage_account=$connection.storage_account
+                --cm_resource_group=$connection.resource_group
+            #else if $connection.provider=="openstack"
+                --co_username=$connection.username
+                --co_password=$connection.password
+                --co_auth_url=$connection.auth_url
+                --co_project_name=$connection.project_name
+                --co_project_domain_name=$connection.project_domain_name
+                --co_user_domain_name=$connection.user_domain_name
+            #end if
+            --bucket=$bucket
+            --object_label=$filename.name.replace(" ", "_")
+            --overwrite_existing=$overwrite_existing
+    </command>
+    <inputs>
+        <param name="filename" type="data" label="Datasets to be downloaded to a cloud-based storage:" />
+        <!--    ************************ NOTE ************************
+                Any change, even order, to the following parameters
+                (provider, provider options, and credential parameters
+                of different providers), MUST be reflected in the
+                CloudManager (lib/galaxy/managers/cloud.py).
+                ******************************************************
+        -->
+        <conditional name="connection">
+            <param name="provider" type="select" label="Cloud-Based Storage Provider">
+                <option value="aws">Amazon Simple Storage Service (S3)</option>
+                <option value="azure">Microsoft Azure Blob Storage</option>
+                <option value="openstack">OpenStack Object Storage (Swift)</option>
+            </param>
+            <when value="aws">
+                <param name="access_key" label="Access key" type="text"/>
+                <param name="secret_key" label="Secret key" type="text"/>
```

Comment:
So when a user clicks view info on these datasets they can see the secret key and such right? These things cannot be tool inputs IMO, am I missing something? 

I think you want to start with a special class of user preferences for these tools (probably based on https://github.com/galaxyproject/galaxy/pull/3383/files) and then add a tool parameter type that takes a credential "name" that the user has configured through their parameters and then use these preferences associated with the user to build the input file? We can then build from there to create per-cloud inputs on the tool form that update the preferences and to make the UX more smooth.

Reply:
@jmchilton thanks for looking into this. I agree it is not a good practice to put user credentials in tool input. Also, the method you linked sounds great. However, we're working on another method that enables accessing cloud-based resource __without__ asking user for their credentials (see #5903 and https://github.com/galaxyproject/cloudauthz, which I'm going to work on it next).